test/dynamo/test_misc.py

```diff
@@ -5070,6 +5066,133 @@ def fn():
         dis.dis(fn)
         self.assertEqual(torch._dynamo.optimize("eager")(fn)(), 3)
 
+    @skipIfNotPy311
+    def test_get_instruction_source_311(self):
+        def f():
+            # flake8: noqa
+            # fmt: off
+            # test binary ops
+            a = ( b   )   +   c
+            a = (a + b) // (c - d)
+            a = b    \
+         +\
+               c  # test
+            a = (
+                (b  # test +
+                    )  \
+                # +
+            << (
+
+                c  # test
+                \
+            )  # test
+            )
+
+            # test slice
+            a = bbb   [  ccc    ]
+            b = bbbbb \
+                [  ccc # test
+
+                 + ddd  \
+
+                ] # test
+            a = bbb[ccc][ddd][eee]
+
+            # test nested and multiline function calls
+            a = g(g(g(b)))
+            a = g(h(
+                g(b),
+                c
+            ))
+
+            # test unicode (match traceback behavior)
+            a = "🔥😂👌" + b
+
+        from torch._dynamo.utils import get_instruction_source_311
+
+        offsets = (3, 11, 15, 19, 23, 29, 35, 46, 58, 64)
+        insts = list(dis.get_instructions(f))
+        expected = (
+            """\
+            a = ( b   )   +   c
+                ~~~~~~~~~~^~~~~
+""",
+            """\
+            a = (a + b) // (c - d)
+                ~~~~~~~~^^~~~~~~~~
+""",
+            """\
+            a = b    \\
+                ~~~~~~
+         +\\
+         ^~
+               c  # test
+               ~
+""",
+            """\
+                (b  # test +
+                ~~~~~~~~~~~~
+                    )  \\
+                    ~~~~
+                # +
+                ~~~
+            << (
+            ^^~~
+
+
+                c  # test
+                ~~~~~~~~~
+                \\
+                ~
+            )  # test
+            ~
+""",
+            """\
+            a = bbb   [  ccc    ]
+                ~~~~~~^^^^^^^^^^^
+""",
+            """\
+            b = bbbbb \\
+                ~~~~~~~
+                [  ccc # test
+                ^^^^^^^^^^^^^
+
+
+                 + ddd  \\
+                 ^^^^^^^^
+
+
+                ] # test
+                ^
+""",
+            """\
+            a = bbb[ccc][ddd][eee]
+                ~~~~~~~~^^^^^
+""",
+            """\
+            a = g(g(g(b)))
+                  ^^^^^^^
+""",
+            """\
+            a = g(h(
+                  ^^
+                g(b),
+                ^^^^^
+                c
+                ^
+            ))
+            ^
+""",
+            """\
+            a = "🔥😂👌" + b
+                ~~~~~~^~~
+""",
```

Comment:
It would be nice to use assertExpectedInline here, which will make it easier to keep these strings up to date.